ocaml/parsing/pprintast.ml

```diff
@@ -1809,18 +1809,20 @@ and extension_constructor_jst _ctxt _f _attrs :
   | _ -> .
 
 and case_list ctxt f l : unit =
-  let aux f {pc_lhs; pc_guard; pc_rhs} =
-    pp f "@;| @[<2>%a%a@;->@;%a@]"
-      (pattern ctxt) pc_lhs (option (guard ctxt) ~first:"@;when@;")
-      pc_guard (expression (under_pipe ctxt)) pc_rhs
+  let aux f {pc_lhs; pc_rhs} =
+    pp f "@;| @[<2>%a@;%a@]" (pattern ctxt) pc_lhs (case_rhs ctxt) pc_rhs
   in
   list aux f l ~sep:""
 
-and guard ctxt f = function
-  | Guard_predicate e -> expression ctxt f e
-  | Guard_pattern { pgp_scrutinee = e; pgp_pattern = pat } ->
-    pp f "@[%a@ match@ %a@]"
-      (expression ctxt) e (pattern ctxt) pat
+and case_rhs ctxt f = function
```

Comment:
Can you confirm that you tested this with `-dsource`? `Psimple_rhs` and `Pboolean_guarded_rhs` look vaguely right (though is `Pboolean_guarded_rhs` missing a `@;` between the `->` and the `%a`) but I'm less sure about `Ppattern_guarded_rhs`.

We could consider dropping the parens if `ppg_cases` is a singleton, but this doesn't actually matter.

Reply:
I investigated this, you can see my preferred version of the formatter in the latest commit.

You may be wondering why we refactor `case_list` inside the new formatter: the reason is that using the existing version of `case_list` gives us some  ugly start and end behavior.

For reference, here is how a single-way match looks before redefining `case_list`:
```
| P1 when e1 match  | P2 -> e2
```
as opposed to after:
```
| P1 when e1 match | P2 -> e2
```

Here is how a multi-way match on one line looks before:
```
| P1 when e1 match ( | P2 -> e2 | P3 -> e3)
```
and after:
```
| P1 when e1 match (| P2 -> e2 | P3 -> e3)
```

Here is how a multi-way match on multiple lines looks before:
```
| P1
      when e1 match (
      | P2 -> e2
      | P3 -> e3
      | P4 -> e4
      | P5 -> e5
      | P6 -> e6)
```
and after:
```
| P1
      when e1 match (
      | P2 -> e2
      | P3 -> e3
      | P4 -> e4
      | P5 -> e5
      | P6 -> e6
      )
```

Let me know what you think!

Reply:
Looks good to me. Could you add this to `ocaml/testsuite/tests/parsetree/source.ml` so future authors are informed when this changes?